EXPERIMENTS.md

```diff
@@ -0,0 +1,188 @@
+ specification in Python, which can be found in [ffg.py]. This specification is
+ using mutually recursive operators, which are not supported by Apalache.  As a
+ result, we are not checking this specification.
+
+ - **Spec 2**. This is the specification [spec/ffg.tla][]. It is a manual
+ adaptation of `Spec 1`. The main difference between `Spec 2` and `Spec 1` is
+ that `Spec 2` uses "folds" (also known as "reduce") instead of recursion. We
+ expect `Spec 2` to be equivalent to `Spec 1`, although we do not have a formal
+ proof of this equivalence. This specification has high model-checking
+ complexity, roughly speaking, as it contains multiple nested folds.
+ 
+ - **Spec 3**. This is the specification [abstract-spec/ffg.tla][]. It is a
+ manual abstraction of `Spec 2` that is highly optimized for constraint solving,
+ especially with Apalache. In addition to that, Spec 3 describes a state machine
+ that adds blocks and votes in every step.
+
+ - **Spec 4**. This is an inductive invariant constructed for `Spec 3`. **TBD**
+ 
+The specifications `Spec 2` and `Spec 3` come with model checking instances
+[spec/MC_ffg.tla] and [abstract-spec/MC_ffg.tla], respectively. These instances
+fix the specification parameters to small values for model checking purposes,
+e.g.:
+
+ - four validators,
+ - from 4 to 10 slots,
+ - up to 12 votes.
+
+By using small values for the parameters, we are able to produce counterexamples
+in a reasonable time when such examples exist. By increasing the parameter
+values, we can increase our confidence in the specification and its properties.
+While this restriction to small parameter values is dictated by the practicality
+of our approach, it is informally supported by the "small scope hypothesis,"
+which was popularized by tools such as [Alloy][].
+
+## 3. All figures in one place
+
+The following table summarizes the experimental figures in one place:
+
+| Experiment | Specification | Property   | Time    | Memory |
+|-----------:|---------------|------------|---------|--------|
+| 4.1        | `Spec 2`      | Row 1      | TBD     | TBD    |
+| 5.1        | `Spec 3`      | Row 2      | TBD     | TBD    |
+| 5.2        | `Spec 3`      | Row 3      | TBD     | TBD    |
+| 6.1        | `Spec 4`      | Row 4      | TBD     | TBD    |
+| 6.2        | `Spec 4`      | Row 5      | TBD     | TBD    |
+
+## 4. Model checking Spec 2
+
+<!-- TODO: Thomas, please add a description of the experiments with MC_ffg_examples.tla -->
+
+### 4.1. Bounded model checking of Accountable Safety
+
+In this experiment, we are checking whether Accountable Safety holds true for
+`Spec 2`. Since `Spec 2` presents a snapshot of a single validator and does not
+make any steps, we have to check this property only in the initial states.
+
+```sh
+$ cd ./spec
+$ JVM_ARGS=-Xmx20G apalache-mc check --length=0 \
+  --inv=AccountableSafety MC_ffg.tla
+```
+
+Note that we have to extend the default JVM heap size from 4G to 20G.
```

Comment:
I'm surprised that you had to do that.
Is it really necessary? I only did it before I added precomputes.

Reply:
yeah, I think it ran out of 4G on my machine, 20G is probably overshooting